### PR TITLE
[docs] Add section 'Dev Container' to the Mojo stdlib docs

### PR DESCRIPTION
**Pull request by benz0li**
Test online with GitHub Codespaces: https://codespaces.new/benz0li/mojo-dev-container?hide_repo_select=true&ref=main  

Parent image: `glcr.b-data.ch/mojo/base:nightly`  
👉 This image is being rebuilt [at the start of every 4th hour](https://cron.help/every-4-hours).

### CODE REVIEW

**Comment by benz0li** [comment]
Cross reference:

* https://github.com/modularml/mojo/discussions/2648

**Comment by JoeLoser** [comment]
Thanks for the contribution!  Can you help me understand some of the rationale on why we'd want to take this change into this repo rather than suggesting you (or someone else) maintain these sort of dev containers from a given published nightly mojo release?  Does it help core contributors to the standard library in some way, or mostly just users of Mojo (not stdlib contributors)?  @gabrieldemarmiesse do you have any thoughts here?

**Comment by benz0li** [comment]
> Does it help core contributors to the standard library in some way, or mostly just users of Mojo (not stdlib contributors)?

For **stdlib contributors**: The `Mojo (nightly)` Dev Container provide a unified development environment.
ℹ️ Currently, `llvm-14` and `llvm-14-tools` are installed in the Dev Containers.
👉 Better v17 with https://apt.llvm.org/llvm.sh as in [.github/workflows/examples.yml](https://github.com/modularml/mojo/blob/db2aa9487617ca4833d766fab9b6f77087894280/.github/workflows/examples.yml)?

For **Mojo users**: In case of an issue, the `Mojo` Dev Container serves as a *reference* [Linux] deployment for reproduction.  
ℹ️ No more "it works on my machine"-discussions.

**Comment by benz0li** [comment]
The [Data Science Dev Containers](https://github.com/b-data/data-science-devcontainers) are meant for Data Scientists, **Mojo users** and alike.

The Dev Containers proposed here are for **core contributors** to the standard library.

And with [GitHub Codespaces](https://docs.github.com/en/codespaces), **anyone** can set up a Mojo development environment.

**Comment by gabrieldemarmiesse** [comment]
@benz0li thank you very much for creating the dev container! @JoeLoser  I think this is a must-have in the future for users and for our contributors' experience. 

While it's extremely easy currently to install the nightly version of Mojo, it might not be as easy to install all the tools needed to run the unit tests (llvm as @benz0li said, or other) and this provides a codespace environment where users can work with Mojo in their browser. That is useful for users who are on platforms not supported by the Mojo sdk or users who want to try Mojo at work but have very tight restrictions on what they can install on their machine. It's also nice as codespace allows you easier reviewing of PRs and quick code modifications without firing your own IDE, adding the remote, checking out the branch, etc...

@benz0li The issue we have right now with the Mojo repository is that the maintainers (specifically, the ones with merging rights) are understaffed. As you can see, a lot of PRs are open, and we have to be very selective with what kind of pull request we accept. We already delayed multiple propositions for new structs and modules, and in my opinion, we should be even more selective with what PRs to accept (we should create a list of new functions that are wanted right now and everything outside of this should be rejected/postponed until we make more progress).

Every piece of code getting into the nightly branch becomes the responsability of the maintainers. Thus if we merge this right now, the maintainers give the garantee to the contributors that the workflow enabled by those 400 lines of code will work all the time, and through updates. We sadly don't have much time to spare for the maintenance. We would need to keep up to date the base image, the llvm install, the list of vscode extensions, etc... Debugging this can take a while too in case of bugs.

My advice would be: Since this is very useful, let's let the contributors know about it through a section in the contributing guide, saying "hey, if you have trouble setting up your environment, just go to https://github.com/b-data/data-science-devcontainers and there are dev containers with everything already installed, and it works with github codespace!". Since this is not our repository, we give no garantee that it works or will be maintained, and for every bug, users will create an issue in https://github.com/b-data/data-science-devcontainers and not in https://github.com/modularml/mojo . When we have more maintainers, and that we have enough bandwidth to maintain a dev workflow in a dev container, then we will bring this PR back to life. Especially since, in the future, I believe it will become easier to install the tools necessary. For example, we'll soon use the Mojo testing framework and I suppose llvm won't be needed! This will simplify the dockerfile :)

**Comment by benz0li** [comment]
> My advice would be: Since this is very useful, let's let the contributors know about it through a section in the contributing guide, saying "hey, if you have trouble setting up your environment, just go to https://github.com/b-data/data-science-devcontainers and there are dev containers with everything already installed, and it works with github codespace!".

I am going to set up dedicated [Mojo Dev Containers](https://github.com/benz0li/mojo-dev-containers), then.

1. Including LLVM v17 (installed using https://apt.llvm.org/llvm.sh)
2. The repository checked out at `~/projects/modularml/mojo`
3. The branch checked out according to the configuration
4. The user's working directory set to the repository

ℹ️ You will be notified here when it is ready.

> Since this is not our repository, we give no garantee that it works or will be maintained, and for every bug, users will create an issue in https://github.com/b-data/data-science-devcontainers and not in https://github.com/modularml/mojo .

I am happy to maintain.

**Comment by gabrieldemarmiesse** [comment]
Great, feel free to modify this PR to add a section to the contributing guide so that new contributors can go to [Mojo Dev Containers](https://github.com/benz0li/mojo-dev-containers) to setup their dev container :)

**Comment by benz0li** [comment]
> Great, feel free to modify this PR to add a section to the contributing guide so that new contributors know they can go to [Mojo Dev Containers](https://github.com/benz0li/mojo-dev-containers) to setup their dev container :)

The [Mojo Dev Container](https://github.com/benz0li/mojo-dev-container) is available now. Please be aware of https://github.com/benz0li/mojo-dev-container/issues/1

I will modify this PR, i.e. add a section to the contributing guide pointing to the [Mojo Dev Container](https://github.com/benz0li/mojo-dev-container).

**Comment by gabrieldemarmiesse** [comment]
> Please be aware of https://github.com/benz0li/mojo-dev-container/issues/1

strange, but it's a good step forward :) 

Thanks for your awesome work, that will hopefully bring Mojo to more users!

File: stdlib/docs/development.md [hunk 1]
```diff
@@ -17,6 +17,22 @@ And if you're using VS Code:
...
+
+The Dev Container also includes the unit test dependencies `lit` and `FileCheck`
+(from LLVM) and has `pre-commit` already set up.
+
+See [Mojo Dev Container &gt; Usage](https://github.com/benz0li/mojo-dev-container#usage)
+on how to use with [Github Codespaces](https://docs.github.com/en/codespaces/developing-in-codespaces/creating-a-codespace-for-a-repository#creating-a-codespace-for-a-repository)
+or [VS Code](https://code.visualstudio.com/docs/devcontainers/containers).
+
+If there is a problem with the Dev Container, please open an issue at
+<https://github.com/benz0li/mojo-dev-container/issues>.
```
Comment (JoeLoser) [inline stdlib/docs/development.md:34]: [here]
```suggestion
[here](https://github.com/benz0li/mojo-dev-container/issues).
```

**Review by JoeLoser — APPROVED** [review]

File: stdlib/docs/development.md [hunk 2]
```diff
@@ -17,6 +17,22 @@ And if you're using VS Code:
...
+prerequisites installed.
+
+The Dev Container also includes the unit test dependencies `lit` and `FileCheck`
+(from LLVM) and has `pre-commit` already set up.
+
+See [Mojo Dev Container &gt; Usage](https://github.com/benz0li/mojo-dev-container#usage)
+on how to use with [Github Codespaces](https://docs.github.com/en/codespaces/developing-in-codespaces/creating-a-codespace-for-a-repository#creating-a-codespace-for-a-repository)
+or [VS Code](https://code.visualstudio.com/docs/devcontainers/containers).
+
+If there is a problem with the Dev Container, please open an issue at
```
Comment (JoeLoser) [inline stdlib/docs/development.md:33]:
```suggestion
If there is a problem with the Dev Container, please open an issue
```

**Review by JoeLoser — APPROVED** [review]
Thanks for writing this up! Happy to see this get incorporated.  I'll check out https://github.com/modularml/mojo/pull/2852 next week as well.

**Comment by JoeLoser** [comment]
!sync

**Pull request closed without merging**